### ISSUE TITLE
Markers in Gaps

### ISSUE THREAD

**Issue by orchetect** [TheAcharya/MarkersExtractor#61]
It's possible to attach markers in gaps. Gaps are empty spaces (no video or audio) between clips on a timeline.

Here are two titles with a gap in between, each with one marker attached:

![clips-and-gap](https://github.com/TheAcharya/MarkersExtractor/assets/11605557/d18ce660-f461-4c59-a6ed-8a3d8329fb77)

Currently in MarkersExtractor, gaps are treated same as video/audio/title clips and gap markers will be included just like any other clip. Final Cut Pro also treats them essentially as "content-less" clips, and they are selectable like any other clip on the timeline. I assume a gap is basically a "generator" for black frames? Or pass-thru (transparent) compositing in the case of layered timelines? Unless there's something more to it than that.

@IAmVigneswaran From your workflow experience, should these be treated differently? Does it matter?

Since there's technically no video, there are no thumbnails that can be generated for markers in gaps. Unless we use a black frame or placeholder frame? That's not hard to do, I'm just not sure what the best option would be.

Would it be of any use to add a CLI flag to ignore markers in gaps?

**Comment by IAmVigneswaran** [comment]
@orchetect 

> Gaps are empty spaces (no video or audio) between clips on a timeline.

Yeap, Gap clips are "Special Clips" to mitigate empty section due to FCP's Magnetic Timeline.

> I assume a gap is basically a "generator" for black frames? Or pass-thru (transparent) compositing in the case of layered timelines? Unless there's something more to it than that.

You can call them as "generator". But there is an Official Black Generator, called Custom.

<img width="643" alt="custom-generator" src="https://github.com/TheAcharya/MarkersExtractor/assets/118706051/646eb454-9493-4589-ad51-75842ea2a9e5">

When you press `Option+W`, you can create a Gap clip anywhere in the timeline. They don't technically have any attributes in the inspector panel.

<img width="424" alt="gap" src="https://github.com/TheAcharya/MarkersExtractor/assets/118706051/85691d79-7fe6-4533-ba72-526fd91376d8">

> From your workflow experience, should these be treated differently? Does it matter?
> 
> Since there's technically no video, there are no thumbnails that can be generated for markers in gaps. Unless we use a black frame or placeholder frame? That's not hard to do, I'm just not sure what the best option would be.

I have seen users (Including myself) who like to place Marker on Gaps clip. This can be for number of reasons.

1. Users would remove the original clip (using the del key), which will result in Gap clips to finesse the timeline.
2. They can place a maker on it. To keep track of shots to be replaced.

To keep our workflow simple. We don't need any additional flags/switch.

1. If there is a Maker on a Gap clip, our tool would read the metadata. 
2. Since it is not a real clip, we print the value of `Gap Clip` under `Clip Name` Column.
3. For `Clip Filename` Column we keep it empty. Similar to `Titles`.
4. We don't need any special black frame? When rendered, FCP does "Flatten" the video when rendering. Apple ProRes 422 does not support transparency.

<img width="902" alt="Export" src="https://github.com/TheAcharya/MarkersExtractor/assets/118706051/f65a2ae5-019e-4142-8b8c-0d911ac1dfeb">

5. That way any Marker on `Gap Clip`will always be extracted as empty black images.

**Comment by orchetect** [comment]
> we print the value of Gap Clip under Clip Name Column

Gap clips can actually be renamed by the user in Final Cut Pro so we'd just use its actual name. Like any other clip, really.

> Apple ProRes 422 does not support transparency.

Yeah I don't mean transparent as in alpha in the final rendered video. I might have to put together a screenshot to explain. Maybe what I'm thinking of isn't even possible in FCP. But I know multiple lanes can exist and maybe overlapping video lanes could happen. In that case, a situation may occur where a gap could overlap with another video. In that case, would the gap be black in that spot or would the gap pass-thru to the video underneath?

**Comment by IAmVigneswaran** [comment]
> Gap clips can actually be renamed by the user in Final Cut Pro so we'd just use its actual name. Like any other clip, really.

<img width="405" alt="Gap-Rename" src="https://github.com/TheAcharya/MarkersExtractor/assets/118706051/44f801f6-e680-4ec5-b77a-d6ebdb609205">

I don't usually rename my Gap clips. Hence, it never crossed my mind. Yes, we just use the actual name. That works. 

> Yeah I don't mean transparent as in alpha in the final rendered video.

Ahh..Noted.

> I might have to put together a screenshot to explain. Maybe what I'm thinking of isn't even possible in FCP. But I know multiple lanes can exist and maybe overlapping video lanes could happen. In that case, a situation may occur where a gap could overlap with another video. In that case, would the gap be black in that spot or would the gap pass-thru to the video underneath?

It is possible in FCP. They are called Secondary Storylines. 

I understand what you mean. I believe this is what you are mentioning about?

**Example 1**
<img width="1233" alt="Gap-Example-SL" src="https://github.com/TheAcharya/MarkersExtractor/assets/118706051/6de89bb9-a1c7-49b7-92d7-f185dcfe3dfc">

Basically the Gap Clip would pass-thru and show the clip from the primary storyline. You can see that the Monitor shows the Pastel clip.

**Example 2**
<img width="1121" alt="Gap-Example-SL2" src="https://github.com/TheAcharya/MarkersExtractor/assets/118706051/f9189836-f087-4a0c-9ebd-c6689a78bb30">

Now the Monitor shows the Green clip. The gap is not in the pass-thru mode.

> Since there's technically no video, there are no thumbnails that can be generated for markers in gaps. Unless we use a black frame or placeholder frame? That's not hard to do, I'm just not sure what the best option would be.

I don't think there is any special pass-thru tag stated in the FCPXML? How are we able to detect the mode that it is in?

**Comment by orchetect** [comment]
> It is possible in FCP. They are called Secondary Storylines.
> the Gap Clip would pass-thru and show the clip from the primary storyline.

Ok great - that's what I was suspecting. Ok I will look into this and see how the XML is structured.

Gaps always pass-thru, don't they? Or are you saying there is actually a mode the user chooses for a gap to be pass-thru or not? From the screenshots it just looks like FCP composites timelines from top down.

**Comment by IAmVigneswaran** [comment]
> Gaps always pass-thru, don't they? Or are you saying there is actually a mode the user chooses for a gap to be pass-thru or not?

Gaps seems to be in always in pass-thru by default. Users can't change it, since it has no clip attributes. But FCP does this automatically base on the "track" layout.

> From the screenshots it just looks like FCP composites timelines from top down.

Yes.

> My guess is that users will want thumbnails to reflect what they see in Final Cut Pro in these scenarios. If we don't handle the pass-thru video behaviour in gaps then they will just see black frames when encountering a gap on the topmost storyline instead.

Yes. In thinking more about it, having a separate switch makes sense. 

`--enable-gap-clips` ?

By default our CLI's behaviour would be pass-thru. When enabled, MarkersExtractor would render a "special image"?

Similar to your build in Audio wave form?

**Comment by orchetect** [comment]
I can't really imagine a scenario where you would want to force-render gaps as blank frames? Seems to me that following FCP's own behavior of compositing the timeline is what users would always want -- as far as thumbnails are concerned.

(For markers, this doesn't actually affect markers. They would still be gathered as normal. Since a marker on a gap would still probably apply to the timeline as a whole, or the user may use the gap as a placeholder to put markers for the storyline(s) underneath.)

**Comment by IAmVigneswaran** [comment]
> I can't really imagine a scenario where you would want to force-render gaps as blank frames?

Yes. Honestly, I can't think of a scenario too. User will always want to see the same behaviour has FCP's. They want show they see thier timeline in the Monitor.

> If there is only one storyline then gap thumbnails would be rendered by MarkersExtractor as black frames since that's the only option we have in that case.

Yes. Your first example.

![clips-and-gap](https://github.com/TheAcharya/MarkersExtractor/assets/11605557/d18ce660-f461-4c59-a6ed-8a3d8329fb77)

> The only argument for having an option to treat gaps differently is if users explicitly want to know in a multi-storyline context that it's the gap that has the marker. But again, I can't really think why that would be super important.

The switch is nice to have. But 99% of all scenario user will leave it has default. Least priority?

**Comment by orchetect** [comment]
> 99% of all scenario user will leave it has default. Least priority?

Exactly. So let's implement the compositing behavior for thumbnails, and not introduce any new options at this time.

**Comment by orchetect** [comment]
Note to self:

To implement gap compositing for thumbnail image generation, one of two solutions is needed:

- A function that walks the XML/data model to generate a "flattened" timeline model. This would generate a series of timecode ranges and their effective corresponding media file, if any. Then it can easily be queried to return the required media file for any given absolute timecode location. And it would be the most performant method, and easier to unit test.
- A function that walks the XML/data model ad-hoc every time a thumbnail is generated. Might be a little slower but wouldn't require a specialized model to perform.

**Comment by IAmVigneswaran** [comment]
An Idea -

**Comment by IAmVigneswaran** [comment]
An Idea,

We can have a "Overlay" Image hard-coded into our tool.

![overlay](https://github.com/TheAcharya/MarkersExtractor/assets/118706051/bf140013-3276-40ed-b7b8-84fea1134c0b)

If the switch is enable,  `--identify-markers-on-gap`

![With-overlay](https://github.com/TheAcharya/MarkersExtractor/assets/118706051/0a2ca3ac-b9d2-45c3-9472-9c0b7fb77df0)

But then again, users can always filter out Gap Clips via Notion or Airtable Database. 

If they did not rename their Gap clips, They can easily created a filter criteria, where column that only contains "Gap" to identify the marker.

It is somewhat becomes redundant.

**Comment by orchetect** [comment]
Yeah I don't think we need anything special for gaps.

As an aside, one of the new capabilities I implemented during the XML parser refactoring is the ability to easily filter out clip types. So if you don't want markers from gaps, or from audio clips, etc. it's very easy.

**Comment by orchetect** [comment]
Sorry, I totally spaced.

We don't need to do any compositing. Users are rendering their video down into a single final video. Thumbnails are just generated based off of that. So whatever FCP does in gaps, that will be rendered in the thumbnail as expected.

We can close this.
